src/extensions/jana/JOmniFactoryGeneratorT.h

```diff
@@ -0,0 +1,65 @@
+// Copyright 2023, Jefferson Science Associates, LLC.
+// Subject to the terms in the LICENSE file found in the top-level directory.
+// Created by Nathan Brei
+
+#pragma once
+
+#include <JANA/JFactorySet.h>
+#include <JANA/JFactoryGenerator.h>
+#include "JOmniFactory.h"
+#include <vector>
+
+template<class FactoryT>
+class JOmniFactoryGeneratorT : public JFactoryGenerator {
+
+public:
+    using FactoryConfigType = typename FactoryT::ConfigType;
+
+    /// Constructor with config
+    explicit JOmniFactoryGeneratorT(std::string tag, std::vector<std::string> default_input_tags, std::vector<std::string> output_tags, FactoryConfigType cfg, JApplication* app):
+            m_tag(tag),
+            m_default_input_tags(std::move(default_input_tags)),
+            m_default_output_tags(std::move(output_tags)),
+            m_default_cfg(cfg),
+            m_app(app)
+    {
+    };
+
+    /// Constructor for NoConfig configuration
+    explicit JOmniFactoryGeneratorT(std::string tag, std::vector<std::string> default_input_tags, std::vector<std::string> output_tags, JApplication* app):
+            m_tag(tag),
+            m_default_input_tags(std::move(default_input_tags)),
+            m_default_output_tags(std::move(output_tags)),
+            m_app(app)
+    {
+    };
+
+    void GenerateFactories(JFactorySet *factory_set) override {
+
+        FactoryT *factory = new FactoryT;
+        factory->SetPluginName(this->GetPluginName());
+        factory->SetFactoryName(JTypeInfo::demangle<FactoryT>());
+        factory->SetTag(m_tag);
+        factory->SetApplication(m_app);
+        factory->GetConfig() = m_default_cfg;
+
+        // Set up all of the wiring prereqs so that Init() can do its thing
+        // Specically, it needs valid input/output tags, a valid logger, and
+        // valid default values in its Config object
+        factory->PreInit(m_tag, m_default_input_tags, m_default_output_tags);
+
+        // We need to call Init() ourselves so that the HelperFactories get created
+        // BEFORE JFactorySet::Add calls GetHelpers().
+        factory->Init();
+
+        // Factory is ready
+        factory_set->Add(factory);
+    }
+
+private:
+    std::string m_tag;
+    std::vector<std::string> m_default_input_tags;
+    std::vector<std::string> m_default_output_tags;
+    FactoryConfigType m_default_cfg;                   /// Default config for a factories. (!) Must be properly copyable
+    JApplication* m_app; // TODO: NWB: Remove me
```

Comment:
We should have jana2 2.1.1 in CI now. I think that has the GetApplication in JMultiFactory which was the only thing that's missing for this, right?

Reply:
Sadly, it looks like we will continue to need to pass m_app to the JOmniFactoryGenerator manually. Why, you ask? For the OmniFactory we need to call Init() before we call JFactorySet::Add, and we can't safely call Init() until the OmniFactory has been given a JApplication. Note this is NOT the case for JChainFactory and JMultifactory, which should be just fine.